### PR TITLE
Application pets

### PR DESCRIPTION
**Pull request by zachholcomb**
Add CRUD functionality for pet applications

### CODE REVIEW

File: app/controllers/application_pets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  def approve
+    pet = Pet.find(params[:pet_id])
+    pet.status = "Pending Adoption"
+    application_pet = pet.find_application_pet(params[:application_id])
+    application_pet.status = "Approved"
+    application_pet.save
+    pet.save
+    redirect_to "/pets/#{params[:pet_id]}"
+  end
```
Comment (Shock253) [inline app/controllers/application_pets_controller.rb]: I wonder if these two approve methods could be more dry?

File: app/controllers/application_pets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  def new
+    @pets = Pet.find(session[:favorites])
+  end
+
+  def create
+    application = Application.create(application_params)
+    if application.save
+      params[:pet_ids].each do |id|
+        application.application_pets.create(pet_id: id)
+      end
```
Comment (zachholcomb) [inline app/controllers/application_pets_controller.rb]: This method got pretty I probably can refactor parts out of it if we have time

File: app/views/application_pets/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+<li><%= @application.phone_number %></li>
+<li><%= @application.description %></li>
+</ul>
+<h2>Pets</h2>
+<% @application.pets.each do |pet| %>
+  <section id="pet-<%= pet.id %>">
+    <%= link_to pet.name, "/pets/#{pet.id}"%>
+    <% if pet.application_approved? %>
+      <p><%= pet.name %> is currently pending adoption.</p>
+      <% if !(@application.pending_approval?(pet.id)) %>
```
Comment (zachholcomb) [inline app/views/application_pets/show.html.erb]: needed a nested if statement to only display revoke link to the application that is approved with that pet

File: config/routes.rb [hunk 4]
```diff
@@ -34,4 +34,13 @@
   delete '/favorites/:pet_id', to: 'favorites#destroy'
   get '/favorites', to: 'favorites#index'
   delete '/favorites', to: 'favorites#delete_all'
+
+  get '/applications/new', to: 'application_pets#new'
+  post '/applications', to: 'application_pets#create'
+  get '/pets/:pet_id/applications', to: 'pets#applications'
+  post '/pets/:pet_id/applications/:application_id', to: 'application_pets#approve'
+  patch '/pets/applications/:application_id', to: 'application_pets#approve_pets'
```
Comment (zachholcomb) [inline config/routes.rb]: wonder if this should be update

File: app/models/application.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class Application < ApplicationRecord
+  validates_presence_of :name, :address, :city, :state, :zip, :phone_number, :description
+  has_many :application_pets, dependent: :destroy
+  has_many :pets, through: :application_pets
+
+  def pending_approval?(pet_id_param)
+    application_pets.where(pet_id: pet_id_param).first.status == "Pending Approval"
```
Comment (Shock253) [inline app/models/application.rb]: to be clear, this method checks if the pet passed in is pending approval on this application

File: app/views/application_pets/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        <p><%= link_to "Revoke #{pet.name}'s Application", "/pets/#{pet.id}/applications/#{@application.id}", method: :delete %></p>
+      <% end %>
+    <% else %>
+      <%= link_to "Approve #{pet.name} for Adoption", "/pets/#{pet.id}/applications/#{@application.id}", method: :post %>
+    <% end %>
+  </section>
+<% end %>
+<h2>Approve Multiple Pets</h2>
```
Comment (zachholcomb) [inline app/views/application_pets/show.html.erb]: this check box cannot be completed if no pets are selected

File: spec/features/applications/applications_show_spec.rb [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    visit "/applications/#{application_2.id}"
+
+    within "#pet-#{@pet_1.id}" do
+      expect(page).to_not have_link("Approve #{@pet_1.name}")
+      expect(page).to have_content("#{@pet_1.name} is currently pending adoption.")
+    end
+
+    click_button "Approve Pets"
+    expect(page).to have_current_path("/applications/#{application_2.id}")
+    expect(page).to have_content("There are currently no pets selected for approval to adopt!")
```
Comment (zachholcomb) [inline spec/features/applications/applications_show_spec.rb]: This flash message shows up if the user clicks submit at the bottom of the page and no pets are checked off, if a pet is already approved its check box will be hidden

**Pull request merged**